drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2017, Texas Instruments Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* The logic here is adapted from SimpleLink SDK's I2CCC32XX.c module. */
+
+#include <kernel.h>
+#include <errno.h>
+#include <i2c.h>
+#include <soc.h>
+#include "i2c-priv.h"
+
+/* Driverlib includes */
+#include <driverlib/rom.h>
+#include <driverlib/rom_map.h>
+#include <driverlib/i2c.h>
+
+#define I2C_MASTER_CMD_BURST_RECEIVE_START_NACK	 I2C_MASTER_CMD_BURST_SEND_START
+#define I2C_MASTER_CMD_BURST_RECEIVE_STOP \
+	I2C_MASTER_CMD_BURST_RECEIVE_ERROR_STOP
+#define I2C_MASTER_CMD_BURST_RECEIVE_CONT_NACK	 I2C_MASTER_CMD_BURST_SEND_CONT
+
+#define DEV_CFG(dev) \
+	((const struct i2c_cc32xx_config *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2c_cc32xx_data *const)(dev)->driver_data)
+#define DEV_BASE(dev) \
+	((DEV_CFG(dev))->base)
+
+
+enum i2c_cc32xx_state {
+	/* I2C is idle, and not performing a transaction */
+	I2C_CC32XX_IDLE_MODE = 0,
+	/* I2C is currently performing a write operation */
+	I2C_CC32XX_WRITE_MODE,
+	/* I2C is currently performing a read operation */
+	I2C_CC32XX_READ_MODE,
+	/* I2C error has occurred */
+	I2C_CC32XX_ERROR = 0xFF
+};
+
+struct i2c_cc32xx_config {
+	u32_t base;
+	u32_t bitrate;
+};
+
+struct i2c_cc32xx_data {
+	struct k_sem mutex;
+	struct k_sem transfer_complete;
+
+	volatile enum i2c_cc32xx_state state;
+
+	u8_t  *write_buf_idx;  /* Internal inc. writeBuf index */
+	size_t write_cnt;      /* Internal dec. writeCounter */
+	u8_t  *read_buf_idx;   /* Internal inc. readBuf index */
+	size_t read_cnt;       /* Internal dec. readCounter */
+	u16_t  slave_addr;     /* Cache slave address for ISR use */
+};
+
+static void configure_i2c_irq(void);
+
+static int i2c_cc32xx_configure(struct device *dev, u32_t dev_config_raw)
+{
+	u32_t base = DEV_BASE(dev);
+	union dev_config dev_config = (union dev_config)dev_config_raw;
+	u32_t bitrate_id;
+
+	if (!dev_config.bits.is_master_device) {
```

Comment:
instead of using bits, can we use shifts & masks of the raw value

Reply:
because ....  these bitfield structures may go away in the future due to the compiler issues discussed in the zephyr mailing list?  OK, best to nip it in the bud.